### PR TITLE
Handle partial results

### PR DESCRIPTION
**Pull request by gdesmar**

### CODE REVIEW

**Comment by cccs-douglass** [comment]
#CybercentreCanada/assemblyline-core/pull/858
#CybercentreCanada/assemblyline-v4-service/pull/657

File: assemblyline/odm/models/config.py [hunk 1]
```diff
@@ -1979,6 +1979,14 @@ class MetadataConfig(odm.Model):
 }
 
 
+TEMPORARY_KEY_TYPE = [
+    # Keep this key as submission wide list merging equal items
+    'union',
+    # Keep this key submission wide on a "las write wins" basis
```
Comment (gdesmar) [inline assemblyline/odm/models/config.py:1985]:
```suggestion
    # Keep this key submission wide on a "last write wins" basis
```

File: assemblyline/odm/models/config.py [hunk 2]
```diff
@@ -2007,8 +2015,15 @@ class Submission(odm.Model):
                              description="Tag types that show up in the submission summary")
     verdicts = odm.Compound(Verdicts, default=DEFAULT_VERDICTS,
                             description="Minimum score value to get the specified verdict.")
+    temporary_keys: dict[str, str] = odm.mapping(odm.enum(TEMPORARY_KEY_TYPE),
```
Comment (gdesmar) [inline assemblyline/odm/models/config.py:2018]:
```suggestion
    temporary_keys: Dict[str, str] = odm.mapping(odm.enum(TEMPORARY_KEY_TYPE),
```

File: assemblyline/odm/models/result.py [hunk 3]
```diff
@@ -179,6 +181,6 @@ def is_empty(self) -> bool:
         if len(self.response.extracted) == 0 and \
                 len(self.response.supplementary) == 0 and \
                 len(self.result.sections) == 0 and \
-                self.result.score == 0:
+                self.result.score == 0 and not self.partial:
```
Comment (gdesmar) [inline assemblyline/odm/models/result.py:184]: Why is partial having an impact on emptiness?

File: assemblyline/odm/models/config.py [hunk 4]
```diff
@@ -2007,8 +2015,15 @@ class Submission(odm.Model):
...
+                                                 description="Set the operation that will be used to update values "
+                                                             "using this key in the temporary submission data.")
 
 
+DEFAULT_TEMPORARY_KEYS = {
+    'passwords': 'union',
```
Comment (gdesmar) [inline assemblyline/odm/models/config.py:2024]:
```suggestion
    'passwords': 'union',
    'email_body': 'overwrite',
```

Comment (gdesmar) [inline assemblyline/odm/models/config.py:2024]:
```suggestion
    'passwords': 'union',
    'email_body': 'union',
```

File: assemblyline/odm/models/result.py (hunk 3 above)
Comment (cccs-douglass) [reply assemblyline/odm/models/result.py:184]: IIRC empty results store and convey _nothing_ other than that a service ran, including whether the result was full or partial.

**Review by cccs-douglass — APPROVED** [review]

**Pull request merged**